### PR TITLE
208004 add scheduled job crds v2

### PR DESCRIPTION
**Pull request by satr**

### CODE REVIEW

File: models/v2/batch_status.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	Name string `json:"name"`
+
+	//Radix batch creation timestamp
+	//
+	// required: true
+	CreationTime string `json:"creationTime"`
+
+	// Status of the Radix batch
+	// required: false
+	Status radixv1.RadixBatchStatus `json:"status,omitempty"`
```
Comment (nilsgstrabo) [inline models/v2/batch_status.go:19]: As discussed: define a model type for status instead of using the CRD type directly

**Review by nilsgstrabo — CHANGES_REQUESTED** [review]
As discussed yesterday; defined specific models instead of using CRDs directly.

File: api/v1/batches/batch_handler.go [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+	CreateBatch(batchScheduleDescription *models.BatchScheduleDescription) (*modelsv1.BatchStatus, error)
+	//MaintainHistoryLimit Delete outdated batches
+	MaintainHistoryLimit() error
+	//DeleteBatch Delete a batch
+	DeleteBatch(batchName string) error
+}
+
+type completedBatchVersioned struct {
+	batchNameV1    string
+	batchNameV2    string
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v1/batches/batch_handler.go]: Fixed

File: api/v1/batches/batch_handler.go [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+	}
+
+	allBatchesPods, err := handler.common.GetPodsForLabelSelector(getLabelSelectorForAllBatchesPods())
+	if err != nil {
+		return nil, err
+	}
+	allBatchesPodsMap := v1.GetPodsToJobNameMap(allBatchesPods)
+	var allRadixBatchStatuses []modelsv1.BatchStatus
+	for idx, batch := range allBatches {
+		allRadixBatchStatuses[idx] = modelsv1.BatchStatus{
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v1/batches/batch_handler.go]: Fixed

File: api/v2/handler.go [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+var (
+	authTransformer mergo.Transformers = mergoutils.CombinedTransformer{Transformers: []mergo.Transformers{mergoutils.BoolPtrTransformer{}}}
+	defaultSrc                         = rand.NewSource(time.Now().UnixNano())
+)
+
+type handler struct {
+	Kube        *kube.Kube
+	Env         *models.Env
+	KubeClient  kubernetes.Interface
+	RadixClient radixclient.Interface
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v2/handler.go]: Fixed

File: api/v1/batches/batch_handler.go [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+				allBatchesPodsMap[batch.Name]),
+		}
+	}
+	log.Debugf("Found %v batches for namespace %s", len(allRadixBatchStatuses), handler.common.Env.RadixDeploymentNamespace)
+	radixBatches, err := handler.common.HandlerApiV2.GetRadixBatches()
+	if err != nil {
+		return nil, err
+	}
+	for _, radixBatch := range radixBatches {
+		radixBatch := radixBatch
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v1/batches/batch_handler.go]: Fixed

File: api/v1/batches/batch_handler.go [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+		return err
+	}
+	batchList, err := handler.getAllBatches()
+	if err != nil {
+		return err
+	}
+
+	completedBatches := convertRadixBatchesToCompletedBatchVersioned(completedRadixBatches.SucceededRadixBatches)
+	log.Debug("maintain history limit for succeeded batches")
+	succeededBatches := batchList.Where(func(j *batchv1.Job) bool { return j.Status.Succeeded > 0 })
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v1/batches/batch_handler.go]: Fixed

File: api/v1/batches/batch_status.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+// GetBatchStatusFromRadixBatch Gets batch status from RadixBatch
+func GetBatchStatusFromRadixBatch(radixBatch *radixv1.RadixBatch) *modelsv1.BatchStatus {
+	jobStatus := modelsv1.BatchStatus{
+		JobStatus: modelsv1.JobStatus{
+			BatchName: radixBatch.GetName(),
+			Created:   utils.FormatTime(&radixBatch.ObjectMeta.CreationTimestamp),
+			Started:   utils.FormatTime(radixBatch.Status.Condition.ActiveTime),
+			Ended:     utils.FormatTime(radixBatch.Status.Condition.CompletionTime),
+			Status:    radixBatch.Status.Condition.Reason, //TODO ?
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v1/batches/batch_status.go]: Fixed

File: api/v1/batches/batch_status.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	modelsv2 "github.com/equinor/radix-job-scheduler/models/v2"
+	radixv1 "github.com/equinor/radix-operator/pkg/apis/radix/v1"
+	v1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// GetBatchStatusFromRadixBatch Gets batch status from RadixBatch
+func GetBatchStatusFromRadixBatch(radixBatch *radixv1.RadixBatch) *modelsv1.BatchStatus {
+	jobStatus := modelsv1.BatchStatus{
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v1/batches/batch_status.go]: Fixed

File: api/v1/jobs/job_handler.go [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+			KubeClient:   kubeClient,
+			RadixClient:  radixClient,
+			Env:          env,
+			HandlerApiV2: apiv2.New(env, kube, kubeClient, radixClient),
+		},
+	}
+}
+
+// GetJobs Get status of all jobs
+func (handler *jobHandler) GetJobs() ([]modelsv1.JobStatus, error) {
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v1/jobs/job_handler.go:65]: Fixed

File: api/v1/jobs/job_handler.go [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+	for idx, k8sJob := range kubeJobs {
+		jobs[idx] = *GetJobStatusFromJob(handler.common.KubeClient, k8sJob, podsMap[k8sJob.Name])
+	}
+
+	log.Debugf("Found %v jobs for namespace %s", len(jobs), handler.common.Env.RadixDeploymentNamespace)
+	return jobs, nil
+}
+
+// GetJob Get status of a job
+func (handler *jobHandler) GetJob(jobName string) (*modelsv1.JobStatus, error) {
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v1/jobs/job_handler.go:100]: Fixed

File: api/v1/jobs/job_handler.go [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+	log.Debugf("create job for namespace: %s", handler.common.Env.RadixDeploymentNamespace)
+	radixBatch, err := handler.common.HandlerApiV2.CreateRadixBatchSingleJob(jobScheduleDescription)
+	if err != nil {
+		return nil, err
+	}
+	return GetJobStatusFromRadixBatchJob(radixBatch), nil
+}
+
+// DeleteJob Delete a job
+func (handler *jobHandler) DeleteJob(jobName string) error {
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v1/jobs/job_handler.go:140]: Fixed

File: api/v1/config_maps.go [hunk 12]
```diff
@@ -1,4 +1,4 @@
-package api
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v1/config_maps.go:1]: Fixed

File: api/v1/owner_references.go [hunk 13]
```diff
@@ -1,4 +1,4 @@
-package api
+package v1
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v1/owner_references.go:1]: Fixed

File: api/v1/services.go [hunk 14]
```diff
@@ -1,9 +1,8 @@
-package api
+package v1
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v1/services.go:1]: Fixed

File: api/v2/handler.go [hunk 15]
```diff
@@ -0,0 +1,571 @@
...
+		return nil, err
+	}
+	var radixBatchJobs []radixv1.RadixBatchJob
+	for _, item := range radixBatchJobWithDescriptions {
+		radixBatchJobs = append(radixBatchJobs, *item.radixBatchJob)
+	}
+	return radixBatchJobs, nil
+}
+
+func creteJobName(radixBatchType kube.RadixBatchType, jobIndex int) string {
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v2/handler.go:496]: Fixed

File: api/v2/handler.go [hunk 16]
```diff
@@ -0,0 +1,571 @@
...
+		return nil, err
+	}
+	return &radixv1.RadixBatchJob{
+		Name:             jobName,
+		JobId:            jobScheduleDescription.JobId,
+		Resources:        jobScheduleDescription.Resources,
+		Node:             jobScheduleDescription.Node,
+		TimeLimitSeconds: jobScheduleDescription.TimeLimitSeconds,
+		BackoffLimit:     jobScheduleDescription.BackoffLimit,
+		Stop:             nil,
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v2/handler.go:588]: Fixed

File: api/v1/secrets.go [hunk 17]
```diff
@@ -1,12 +1,12 @@
-package api
+package v1
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v1/secrets.go:1]: Fixed

File: api/v2/handler.go [hunk 18]
```diff
@@ -0,0 +1,571 @@
...
+}
+
+func (h *handler) createSecrets(namespace string, secrets []*corev1.Secret) error {
+	var errs []error
+	for _, secret := range secrets {
+		if secret.Data == nil || len(secret.Data) == 0 {
+			continue //if Data is empty, the secret is not used in any jobs
+		}
+		_, err := h.Kube.KubeClient().CoreV1().Secrets(namespace).Create(context.TODO(), secret, metav1.CreateOptions{})
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v2/handler.go:554]: Fixed

File: models/v1/defaults/batch.go [hunk 19]
```diff
@@ -0,0 +1,17 @@
+package defaultsv1
```
[the remark this answers is not in the document]
Comment (satr) [reply models/v1/defaults/batch.go:1]: Fixed

File: models/v1/defaults/job.go [hunk 20]
```diff
@@ -1,4 +1,4 @@
-package defaults
+package defaultsv1
```
[the remark this answers is not in the document]
Comment (satr) [reply models/v1/defaults/job.go:1]: Fixed

File: models/v1/job_list.go [hunk 21]
```diff
@@ -1,4 +1,4 @@
-package models
+package v1
```
[the remark this answers is not in the document]
Comment (satr) [reply models/v1/job_list.go]: Fixed

File: models/v2/batch_status.go (hunk 1 above)
Comment (satr) [reply models/v2/batch_status.go]: Fixed

File: api/v1/labels.go [hunk 22]
```diff
@@ -1,4 +1,4 @@
-package api
+package v1
```
[the remark this answers is not in the document]
Comment (satr) [reply api/v1/labels.go:1]: Fixed

File: api/v1/services.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (nilsgstrabo) [reply api/v1/services.go:1]: Method CreateService is not used

[the remark this answers is not in the document]
Comment (satr) [reply api/v1/services.go:1]: Removed

**Review by nilsgstrabo — APPROVED** [review]

**Pull request merged**